bug - codegen: internal module imports missing crate:: prefix (multi-file projects fail to build)

### Area
- Compiler (frontend/backend/codegen)
- Tooling (CLI/formatter/test runner)

### Summary
Expected: In multi-file projects, generated Rust should reference sibling modules using `crate::...` (or `super::...`) so module imports compile correctly.
Actual: Generated Rust submodules emit imports like `use db::schema::Database;` instead of `use crate::db::schema::Database;`, causing Rust 2018+ to look for external crate `db` and fail.

### Reproduction steps
1) Create a multi-file Incan project:

```
src/
  main.incn
  db/
    schema.incn
  store/
    json_store.incn
```

2) In `store/json_store.incn`, import from the sibling module:

```incan
from db.schema import Database
```

3) Build:
`incan build src/main.incn`

4) Observe Rust compilation failure in the generated project:
- generated `src/store/json_store.rs` contains `use db::schema::Database;`
- Rust error: unresolved module or unlinked crate `db`

### Output / logs
```text
error[E0433]: failed to resolve: use of unresolved module or unlinked crate `db`
 --> src/store/json_store.rs:...
  |
  | use db::schema::Database;
  |     ^^ use of unresolved module or unlinked crate `db`
```

### Environment
OS: macOS (darwin)
Rust: unknown
Incan: v0.1.0-beta.2 (observed via generated header)
Command: `incan build <entry>.incn`

--- a/src/backend/ir/codegen.rs
+++ b/src/backend/ir/codegen.rs
@@ -371,11 +371,15 @@ impl<'a> IrCodegen<'a> {
         }
 
         // Use the IR pipeline: AST → IR → Rust
-        self.try_generate_via_ir(program)
+        self.try_generate_via_ir(program, &HashSet::new())
     }
 
     /// Generate code via the IR pipeline (fallible version)
-    fn try_generate_via_ir(&self, program: &Program) -> Result<String, GenerationError> {
+    fn try_generate_via_ir(
+        &self,
+        program: &Program,
+        internal_module_roots: &HashSet<String>,
+    ) -> Result<String, GenerationError> {
         // Typecheck to obtain reusable type information for lowering.
         //
         // Strict policy: if typechecking fails, do NOT proceed to lowering/codegen.
@@ -413,6 +417,7 @@ impl<'a> IrCodegen<'a> {
             let mut svc = EmitService::new_from_program(&ir_program);
             // Configure inner emitter
             let inner = svc.inner_mut();
+            inner.set_internal_module_roots(internal_module_roots.clone());
             if self.emit_zen_in_main {
                 inner.set_emit_zen(true);
             }
@@ -424,6 +429,7 @@ impl<'a> IrCodegen<'a> {
             Ok(svc.emit_program(&ir_program)?)
         } else {
             let mut emitter = IrEmitter::new(&unified_registry);
+            emitter.set_internal_module_roots(internal_module_roots.clone());
             if self.emit_zen_in_main {
                 emitter.set_emit_zen(true);
             }
@@ -458,12 +464,22 @@ impl<'a> IrCodegen<'a> {
         let mut lowering = AstLowering::new();
         let ir_program = lowering.lower_program(program)?;
 
+        // Best-effort: treat registered dependency module names as internal roots.
+        // (This is most relevant for the non-nested multi-file API.)
+        let internal_roots: HashSet<String> = self
+            .dependency_modules
+            .iter()
+            .map(|(name, _)| (*name).to_string())
+            .collect();
+
         let use_emit_service = env::var("INCAN_EMIT_SERVICE").ok().as_deref() == Some("1");
         if use_emit_service {
             let mut svc = EmitService::new_from_program(&ir_program);
+            svc.inner_mut().set_internal_module_roots(internal_roots);
             Ok(svc.emit_program(&ir_program)?)
         } else {
             let mut emitter = IrEmitter::new(&ir_program.function_registry);
+            emitter.set_internal_module_roots(internal_roots);
             if self.emit_zen_in_main {
                 emitter.set_emit_zen(true);
             }
@@ -527,8 +543,10 @@ impl<'a> IrCodegen<'a> {
             self.collect_rust_crates(dep_ast);
         }
 
+        let internal_roots: HashSet<String> = module_names.iter().map(|s| (*s).to_string()).collect();
+
         // Generate main file
-        let main_code = self.try_generate_via_ir(program)?;
+        let main_code = self.try_generate_via_ir(program, &internal_roots)?;
 
         // Generate module files
         let mut modules = HashMap::new();
@@ -539,9 +557,11 @@ impl<'a> IrCodegen<'a> {
                 let use_emit_service = env::var("INCAN_EMIT_SERVICE").ok().as_deref() == Some("1");
                 let module_code = if use_emit_service {
                     let mut svc = EmitService::new_from_program(&ir);
+                    svc.inner_mut().set_internal_module_roots(internal_roots.clone());
                     svc.emit_program(&ir)?
                 } else {
                     let mut emitter = IrEmitter::new(&ir.function_registry);
+                    emitter.set_internal_module_roots(internal_roots.clone());
                     emitter.emit_program(&ir)?
                 };
                 modules.insert(name.to_string(), module_code);
@@ -604,8 +624,10 @@ impl<'a> IrCodegen<'a> {
             self.collect_rust_crates(dep_ast);
         }
 
+        let internal_roots: HashSet<String> = module_paths.iter().filter_map(|p| p.first().cloned()).collect();
+
         // Generate main file
-        let main_code = self.try_generate_via_ir(program)?;
+        let main_code = self.try_generate_via_ir(program, &internal_roots)?;
 
         // Generate module files by path
         let mut modules = HashMap::new();
@@ -620,9 +642,11 @@ impl<'a> IrCodegen<'a> {
                     let use_emit_service = env::var("INCAN_EMIT_SERVICE").ok().as_deref() == Some("1");
                     let module_code = if use_emit_service {
                         let mut svc = EmitService::new_from_program(&ir);
+                        svc.inner_mut().set_internal_module_roots(internal_roots.clone());
                         svc.emit_program(&ir)?
                     } else {
                         let mut emitter = IrEmitter::new(&ir.function_registry);
+                        emitter.set_internal_module_roots(internal_roots.clone());
                         emitter.emit_program(&ir)?
                     };
                     modules.insert(path.clone(), module_code);
@@ -651,6 +675,64 @@ mod tests {
         let tokens = lexer::lex(source).unwrap();
         let ast = parser::parse(&tokens).unwrap();
         IrCodegen::new().try_generate(&ast).unwrap()
+    }
+
+    /// Parse an Incan program into an AST
+    fn parse_program(source: &str) -> Program {
+        let tokens = lexer::lex(source).unwrap();
+        parser::parse(&tokens).unwrap()
+    }
+
+    fn db_module_program() -> Program {
+        parse_program(
+            r#"
+model Database:
+  id: int
+"#,
+        )
+    }
+
+    fn main_module_program() -> Program {
+        parse_program(
+            r#"
+def main() -> None:
+  return
+"#,
+        )
+    }
+
+    fn generate_nested_store_code(store_source: &str) -> String {
+        let db_module = db_module_program();
+        let store_module = parse_program(store_source);
+        let main_module = main_module_program();
+
+        let mut codegen = IrCodegen::new();
+        codegen.add_module("db_schema", &db_module);
+        codegen.add_module("store_json_store", &store_module);
+
+        let db_path = vec!["db".to_string(), "schema".to_string()];
+        let store_path = vec!["store".to_string(), "json_store".to_string()];
+        let (_main_code, rust_modules) = codegen
+            .try_generate_multi_file_nested(&main_module, &[db_path.clone(), store_path.clone()])
+            .unwrap();
+
+        rust_modules.get(&store_path).unwrap().to_string()
+    }
+
+    fn generate_non_nested_store_code(store_source: &str, db_module_name: &str) -> String {
+        let db_module = db_module_program();
+        let store_module = parse_program(store_source);
+        let main_module = main_module_program();
+
+        let mut codegen = IrCodegen::new();
+        codegen.add_module(db_module_name, &db_module);
+        codegen.add_module("store", &store_module);
+
+        let (_main_code, modules) = codegen
+            .try_generate_multi_file(&main_module, &[db_module_name, "store"])
+            .unwrap();
+
+        modules.get("store").unwrap().to_string()
     }
 
     #[test]
@@ -801,5 +883,84 @@ enum Status:
         assert!(code.contains("enum Status"));
         assert!(code.contains("Active"));
         assert!(code.contains("Inactive"));
+    }
+
+    #[test]
+    fn test_multi_file_imports_use_crate_prefix() {
+        let store_code = generate_nested_store_code(
+            r#"
+from db.schema import Database
+"#,
+        );
+        assert!(store_code.contains("use crate::db::schema::Database;"));
+        assert!(!store_code.contains("use db::schema::Database;"));
+    }
+
+    #[test]
+    fn test_rust_imports_do_not_use_crate_prefix() {
+        let code = generate(
+            r#"
+from rust::time import Duration
+"#,
+        );
+        assert!(code.contains("use time::Duration;"));
+        assert!(!code.contains("use crate::time::Duration;"));
+    }
+
+    #[test]
+    fn test_rust_style_external_crate_import_is_not_forced_under_crate() {
+        let code = generate(
+            r#"
+import serde::Serialize
+"#,
+        );
+        assert!(code.contains("use serde::Serialize;"));
+        assert!(!code.contains("use crate::serde::Serialize;"));
+    }
+
+    #[test]
+    fn test_relative_from_import_uses_super_prefix() {
+        let store_code = generate_nested_store_code(
+            r#"
+from ..db.schema import Database
+"#,
+        );
+        assert!(store_code.contains("use super::db::schema::Database;"));
+        assert!(!store_code.contains("use crate::db::schema::Database;"));
+    }
+
+    #[test]
+    fn test_multi_file_imports_rust_style_module_import_uses_crate_prefix() {
+        let store_code = generate_nested_store_code(
+            r#"
+import db::schema::Database
+"#,
+        );
+        assert!(store_code.contains("use crate::db::schema::Database;"));
+        assert!(!store_code.contains("use db::schema::Database;"));
+    }
+
+    #[test]
+    fn test_non_nested_multi_file_api_sets_internal_module_roots() {
+        let store_code = generate_non_nested_store_code(
+            r#"
+from db import Database
+"#,
+            "db",
+        );
+        assert!(store_code.contains("use crate::db::Database;"));
+        assert!(!store_code.contains("use db::Database;"));
+    }
+
+    #[test]
+    fn test_non_nested_multi_file_nested_modules_use_crate_prefix() {
+        let store_code = generate_non_nested_store_code(
+            r#"
+from db.schema import Database
+"#,
+            "db_schema",
+        );
+        assert!(store_code.contains("use crate::db::schema::Database;"));
+        assert!(!store_code.contains("use db::schema::Database;"));
     }
 }

--- a/src/backend/ir/decl.rs
+++ b/src/backend/ir/decl.rs
@@ -51,6 +51,7 @@ pub enum IrDeclKind {
 
     /// Import (preserved for codegen)
     Import {
+        qualifier: IrImportQualifier,
         path: Vec<String>,
         alias: Option<String>,
         /// Specific items being imported (for `from x import a, b`)
@@ -59,6 +60,32 @@ pub enum IrDeclKind {
 
     /// Impl block for methods on structs/enums
     Impl(IrImpl),
+}
+
+/// How an import path should be qualified in generated Rust.
+///
+/// ## Background (why this exists)
+/// In Rust 2018+ module paths in `use ...` are **not implicitly crate-rooted** when emitted inside a submodule.
+/// For example, inside `store::json_store`, `use db::schema::Database;` resolves as `store::json_store::db::...`
+/// (or an external crate), not `crate::db::...`. For multi-file Incan projects this commonly needs an explicit
+/// `crate::` (or `super::`) prefix for correctness.
+///
+/// We preserve the required qualification intent in IR so codegen can emit correct `use` paths.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum IrImportQualifier {
+    /// No qualifier (external crate or special-case import).
+    None,
+    /// Decide at emit-time whether this should be `crate::...` or unqualified.
+    ///
+    /// This is used to avoid semantic regressions: `import serde::Serialize` should remain an external crate import
+    /// unless `serde` is a known internal module root in the current compilation unit.
+    ///
+    /// The emitter uses the set of known internal module roots (for multi-file builds) to decide whether to prefix.
+    Auto,
+    /// Prefix with `crate::` (absolute import in the current crate).
+    Crate,
+    /// Prefix with `super::` repeated N times (relative import).
+    Super(usize),
 }
 
 /// An item in a from ... import statement

--- a/src/backend/ir/emit/decls.rs
+++ b/src/backend/ir/emit/decls.rs
@@ -15,7 +15,7 @@
 
 use std::collections::HashSet;
 
-use proc_macro2::{Ident, Literal, TokenStream};
+use proc_macro2::{Literal, TokenStream};
 use quote::{format_ident, quote};
 
 use incan_core::lang::conventions;
@@ -25,19 +25,19 @@ use incan_core::lang::stdlib;
 
 const ZEN_TEXT: &str = include_str!("../../../../stdlib/zen.txt");
 
-use super::super::decl::{IrDecl, IrDeclKind};
+use super::super::decl::{IrDecl, IrDeclKind, IrImportQualifier};
 use super::super::expr::{IrExpr, IrExprKind, MethodKind, VarAccess};
 use super::super::stmt::{AssignTarget, IrStmt, IrStmtKind};
 use super::super::types::IrType;
 use super::{EmitError, IrEmitter};
 
-fn join_path_idents(segments: &[Ident]) -> TokenStream {
+fn join_path_tokens(segments: &[TokenStream]) -> TokenStream {
     let mut ts = TokenStream::new();
     for (idx, seg) in segments.iter().enumerate() {
         if idx > 0 {
             ts.extend(quote! { :: });
         }
-        ts.extend(quote! { #seg });
+        ts.extend(seg.clone());
     }
     ts
 }
@@ -421,7 +421,12 @@ impl<'a> IrEmitter<'a> {
                     #vis const #name_ident: #ty_tokens = #value_tokens;
                 })
             }
-            IrDeclKind::Import { path, alias, items } => {
+            IrDeclKind::Import {
+                qualifier,
+                path,
+                alias,
+                items,
+            } => {
                 // Skip serde imports if we're already importing them automatically
                 if self.needs_serde && path.len() == 1 && path[0] == "serde" {
                     let is_serde_trait = items.iter().any(|item| {
@@ -440,14 +445,38 @@ impl<'a> IrEmitter<'a> {
                 // - `testing` maps to `incan_stdlib::testing`
                 let is_stdlib_web = path.first().map(|s| s == stdlib::STDLIB_WEB).unwrap_or(false);
                 let is_stdlib_testing = path.first().map(|s| s == stdlib::STDLIB_TESTING).unwrap_or(false);
-                let path_tokens: Vec<_> = if is_stdlib_web {
-                    vec![format_ident!("incan_stdlib"), format_ident!("web")]
+                let mapped_path_tokens: Vec<_> = if is_stdlib_web {
+                    vec![quote! { incan_stdlib }, quote! { web }]
                 } else if is_stdlib_testing {
-                    vec![format_ident!("incan_stdlib"), format_ident!("testing")]
+                    vec![quote! { incan_stdlib }, quote! { testing }]
                 } else {
-                    path.iter().map(|s| format_ident!("{}", s)).collect()
+                    path.iter()
+                        .map(|s| {
+                            let ident = format_ident!("{}", s);
+                            quote! { #ident }
+                        })
+                        .collect()
                 };
-                let path_ts = join_path_idents(&path_tokens);
+                let mut path_tokens: Vec<TokenStream> = Vec::new();
+                let apply_prefix = !(is_stdlib_web || is_stdlib_testing);
+                if apply_prefix {
+                    match qualifier {
+                        IrImportQualifier::Auto => {
+                            if self.is_internal_module_path(path) {
+                                path_tokens.push(quote! { crate });
+                            }
+                        }
+                        IrImportQualifier::Crate => path_tokens.push(quote! { crate }),
+                        IrImportQualifier::Super(levels) => {
+                            for _ in 0..*levels {
+                                path_tokens.push(quote! { super });
+                            }
+                        }
+                        IrImportQualifier::None => {}
+                    }
+                }
+                path_tokens.extend(mapped_path_tokens);
+                let path_ts = join_path_tokens(&path_tokens);
 
                 if let Some(alias_name) = alias {
                     let alias_ident = format_ident!("{}", alias_name);
@@ -460,7 +489,7 @@ impl<'a> IrEmitter<'a> {
                         .map(|item| {
                             let name_ident = format_ident!("{}", &item.name);
                             let path_tokens_clone = path_tokens.clone();
-                            let path_ts_clone = join_path_idents(&path_tokens_clone);
+                            let path_ts_clone = join_path_tokens(&path_tokens_clone);
                             if let Some(alias) = &item.alias {
                                 let alias_ident = format_ident!("{}", alias);
                                 quote! { use #path_ts_clone :: #name_ident as #alias_ident; }
@@ -470,7 +499,7 @@ impl<'a> IrEmitter<'a> {
                         })
                         .collect();
                     Ok(quote! { #(#item_stmts)* })
-                } else if path_tokens.len() == 1 {
+                } else if path.len() == 1 && !is_stdlib_web && !is_stdlib_testing {
                     Ok(quote! {})
                 } else {
                     Ok(quote! {

--- a/src/backend/ir/emit/mod.rs
+++ b/src/backend/ir/emit/mod.rs
@@ -28,6 +28,7 @@ mod types;
 pub use errors::EmitError;
 
 use std::cell::RefCell;
+use std::collections::HashSet;
 
 use super::FunctionRegistry;
 use super::decl::VariantFields;
@@ -77,6 +78,10 @@ pub struct IrEmitter<'a> {
     const_string_literals: std::collections::HashMap<String, String>,
     /// Collected routes for web emission
     routes: Vec<RouteSpec>,
+    /// Known internal module roots for this compilation unit (e.g. {"db", "store"}).
+    ///
+    /// Used to disambiguate crate-internal module imports vs external crate imports when emitting `use` paths.
+    internal_module_roots: HashSet<String>,
 }
 
 impl<'a> IrEmitter<'a> {
@@ -102,7 +107,32 @@ impl<'a> IrEmitter<'a> {
             in_return_context: RefCell::new(false),
             const_string_literals: std::collections::HashMap::new(),
             routes: Vec::new(),
+            internal_module_roots: HashSet::new(),
         }
+    }
+
+    /// Set the internal module roots (top-level module names) for a multi-file compilation.
+    pub fn set_internal_module_roots(&mut self, roots: HashSet<String>) {
+        self.internal_module_roots = roots;
+    }
+
+    /// Check if a top-level name is a known internal module root.
+    pub(crate) fn is_internal_module_root(&self, name: &str) -> bool {
+        self.internal_module_roots.contains(name)
+    }
+
+    /// Check if a full module path is known internally.
+    pub(crate) fn is_internal_module_path(&self, segments: &[String]) -> bool {
+        if let Some(first) = segments.first() {
+            if self.is_internal_module_root(first) {
+                return true;
+            }
+        }
+        if segments.is_empty() {
+            return false;
+        }
+        let joined = segments.join("_");
+        self.internal_module_roots.contains(&joined)
     }
 
     /// Set external rust functions.

--- a/src/backend/ir/lower/decl.rs
+++ b/src/backend/ir/lower/decl.rs
@@ -765,6 +765,28 @@ impl AstLowering {
             ast::ImportKind::Python(s) => (vec![s.clone()], vec![]),
         };
 
+        let qualifier = match &i.kind {
+            ast::ImportKind::Module(p) => {
+                if p.parent_levels > 0 {
+                    super::super::decl::IrImportQualifier::Super(p.parent_levels)
+                } else if p.is_absolute {
+                    super::super::decl::IrImportQualifier::Crate
+                } else {
+                    super::super::decl::IrImportQualifier::Auto
+                }
+            }
+            ast::ImportKind::From { module, .. } => {
+                if module.parent_levels > 0 {
+                    super::super::decl::IrImportQualifier::Super(module.parent_levels)
+                } else if module.is_absolute {
+                    super::super::decl::IrImportQualifier::Crate
+                } else {
+                    super::super::decl::IrImportQualifier::Auto
+                }
+            }
+            _ => super::super::decl::IrImportQualifier::None,
+        };
+
         // Convert AST import items to IR import items
         let ir_items: Vec<super::super::decl::IrImportItem> = ast_items
             .iter()
@@ -775,6 +797,7 @@ impl AstLowering {
             .collect();
 
         IrDeclKind::Import {
+            qualifier,
             path,
             alias: i.alias.clone(),
             items: ir_items,

--- a/workspaces/docs-site/docs/language/how-to/rust_interop.md
+++ b/workspaces/docs-site/docs/language/how-to/rust_interop.md
@@ -18,6 +18,14 @@ from rust::std::collections import HashMap, HashSet
 import rust::serde_json::Value
 ```
 
+### Note on Rust-style imports without `rust::`
+
+Incan also supports Rust-style module paths in regular imports (e.g. `import std::fs`).
+
+- For the Rust standard library, `std::...` is supported as a convenience.
+- For crates from crates.io (e.g. `serde`, `regex`, `reqwest`), prefer `rust::...` so the compiler can manage
+  dependencies in the generated `Cargo.toml`.
+
 ## Automatic Dependency Management
 
 When you use `import rust::crate_name`, Incan automatically adds the dependency to your generated `Cargo.toml`.

--- a/workspaces/docs-site/docs/language/reference/imports_and_modules.md
+++ b/workspaces/docs-site/docs/language/reference/imports_and_modules.md
@@ -191,7 +191,7 @@ def main() -> None:
 
 ## Rust standard library access
 
-Incan can import from Rust’s standard library:
+Incan can import from Rust’s standard library.
 
 ```incan
 import std::fs
@@ -200,8 +200,22 @@ import std::path::Path
 import std::time
 ```
 
+This is equivalent to using the explicit Rust interop prefix:
+
+```incan
+import rust::std::fs
+```
+
 Note: using these requires understanding the underlying Rust types. Prefer Incan built-ins (`read_file`, `write_file`,
 etc.) where available.
+
+## Rust crates vs Incan modules (important)
+
+- **External crates**: Prefer `rust::...` imports (e.g. `import rust::serde_json`), which also enables automatic
+  dependency management for generated `Cargo.toml`.
+- **Incan project modules** (multi-file projects): imports like `from db.schema import Database` refer to modules in
+  the current crate and are emitted as `crate::db::schema::Database` in generated Rust so they compile reliably from
+  submodules.
 
 ## Current status and limitations
 

